File: core/Stock.js
import { WebullBot } from "./WebullBot";
import colors from "colors";

export class Stock {
  ticker = "";
  amount = 0;
  currentPrice = 1;
  prevPrice = 1;
  change = 0;
  buyPrice = 0;
  sellPrice = 0;
  profit = 0;
  tickNum = 0;
  direction = "same";
  avaialableFunds = 0;
  percentChangeSellPositive = 0.02;
  percentChangeSellNegative = -0.02;

  constructor(ticker, wb) {
    this.ticker = ticker;
    this.getPrice(wb);
  }

  getPrice(wb) {
    const tick = wb.ticker(this.ticker);
    this.prevPrice = this.currentPrice;
    this.currentPrice = tick["price"];
  }

  tick(wb, bank) {
    this.tickNum = this.tickNum + 1;
    this.getPrice(wb);
    this.change = Math.round(this.currentPrice - this.prevPrice);
    if (this.change >= this.percentChangeSellPositive) this.direction = "up";
    else if (this.change < this.percentChangeSellNegative)
      this.direction = "down";
    else this.direction = "same";

    //console.log(`P: ${this.prevPrice} C: ${this.currentPrice} Ch: ${this.change} D: ${this.direction}`)
    this.considerBuy(bank);
    this.considerSell();
  }

  considerBuy(bank) {
    if (this.buyPrice == 0 && this.direction == "up") {
      this.avaialableFunds = bank.get_money(this.ticker, this.currentPrice);
      if (this.avaialableFunds > this.currentPrice) {
        this.buy(this.avaialableFunds / this.currentPrice);
        this.direction = "same";
      }
    }
  }

  considerSell() {
    if (this.buyPrice != 0 && this.sellPrice == 0)
      if (
        this.currentPrice <= this.buyPrice + this.percentChangeSellNegative ||
        this.currentPrice >= this.buyPrice + this.percentChangeSellPositive
      )
        this.sell();
  }

  buy(amount) {
    console.log(
      `Buying {amount} shares of {this.ticker} at {this.currentPrice} each`.cyan
    );
    this.buyPrice = this.currentPrice;
    this.avaialableFunds -= amount * this.buyPrice;
    this.amount = amount;
  }

  sell() {
    this.sellPrice = this.currentPrice;
    this.avaialableFunds += this.amount * this.sellPrice;
    this.profit = Math.round((this.sellPrice - this.buyPrice) * this.amount);
    if (this.profit < 0)
      console.log(
        `Selling ${this.amount} shares of ${this.ticker} at ${this.currentPrice} each Buy Price: ${this.buyPrice} Profit: ${this.profit}`
          .red
      );
    else
      console.log(
        `Selling ${this.amount} shares of ${this.ticker} at ${this.currentPrice} each Buy Price: ${this.buyPrice} Profit: ${this.profit}`
          .green
      );
  }

  reset() {
    this.amount = 0;
    this.currentPrice = 1;
    this.prevPrice = 1;
    this.change = 0;
    this.buyPrice = 0;
    this.sellPrice = 0;
    this.profit = 0;
    this.tickNum = 0;
    this.direction = "same";
    this.avaialableFunds = 0;
  }
}

File: server/core/Stock.py
from core.WebullBot import WebullBot
from termcolor import colored

class Stock:
    ticker: str = ''
    amount: int = 0
    currentPrice: float = 1
    prevPrice: float = 1
    change: float = 0
    buyPrice: float = 0
    sellPrice: float = 0
    profit: float = 0
    tickNum: int = 0
    direction: str = 'same'
    avaialableFunds: int = 0
    percentChangeSellPositive: float = 0.02
    percentChangeSellNegative: float = -0.02

    def __init__(self, ticker: str,  wb: WebullBot):
        self.ticker = ticker
        self.getPrice(wb)
    
    def getPrice(self, wb: WebullBot):
        tick = wb.ticker(self.ticker)
        self.prevPrice = self.currentPrice 
        self.currentPrice = tick['price']
    
    def tick(self, wb: WebullBot, bank):
        self.tickNum = self.tickNum + 1
        self.getPrice(wb)       
        self.change = round(self.currentPrice - self.prevPrice , 2)
        if self.change >= self.percentChangeSellPositive:
            self.direction = 'up'
        elif self.change < self.percentChangeSellNegative:
            self.direction = 'down'
        else:   
            self.direction = 'same'
        
        print(f'P: {self.prevPrice} C: {self.currentPrice} Ch: {self.change} D: {self.direction}') if self.tickNum % 30 == 0 else None
        self.considerBuy(bank)
        self.considerSell()

        
    def considerBuy(self, bank):
        if self.buyPrice == 0 and self.direction == 'up':
            self.avaialableFunds = bank.get_money(self.ticker, self.currentPrice)
            if self.avaialableFunds > self.currentPrice:
                self.buy(self.avaialableFunds // self.currentPrice)
                self.direction = 'same'
    
    def considerSell(self):
        if self.buyPrice != 0 and self.sellPrice == 0:
            if self.currentPrice <= self.buyPrice + self.percentChangeSellNegative or self.currentPrice >= self.buyPrice + self.percentChangeSellPositive:
                self.sell()
    
    def buy(self, amount: int):
        print(colored(f'Buying {amount} shares of {self.ticker} at {self.currentPrice} each', 'cyan'))
        self.buyPrice = self.currentPrice
        self.avaialableFunds -= amount * self.buyPrice
        self.amount = amount


    def sell(self):
        self.sellPrice = self.currentPrice
        self.avaialableFunds += self.amount * self.sellPrice
        self.profit = round((self.sellPrice - self.buyPrice) * self.amount, 2)
        print(colored(f'Selling {self.amount} shares of {self.ticker} at {self.currentPrice} each Buy Price: {self.buyPrice} Profit: {self.profit}',  'red' if self.profit < 0 else 'green'))
       
    def reset(self):
        self.amount = 0
        self.currentPrice = 1
        self.prevPrice = 1
        self.change = 0
        self.buyPrice = 0
        self.sellPrice = 0
        self.profit = 0
        self.tickNum = 0
        self.direction = 'same'
        self.avaialableFunds = 0